File: editor/test/editor/progress_test.clj
;; Copyright 2020-2026 The Defold Foundation
;; Copyright 2014-2020 King
;; Copyright 2009-2014 Ragnar Svensson, Christian Murray
;; Licensed under the Defold License version 1.0 (the "License"); you may not use
;; this file except in compliance with the License.
;;
;; You may obtain a copy of the License, together with FAQs at
;; https://www.defold.com/license
;;
;; Unless required by applicable law or agreed to in writing, software distributed
;; under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
;; specific language governing permissions and limitations under the License.

(ns editor.progress-test
  (:refer-clojure :exclude [mapv])
  (:require [clojure.test :refer :all]
            [editor.localization :as localization]
            [editor.progress :as progress])
  (:import [editor.progress Progress]))

(set! *warn-on-reflection* true)
(set! *unchecked-math* :warn-on-boxed)

(defn- to-map [^Progress progress]
  (into {} progress))

(deftest make-test
  (are [expected-fields progress]
    (let [actual-fields (to-map progress)]
      (is (instance? Progress progress))
      (is (= expected-fields actual-fields)))

    {:message (localization/message "a") :size 1 :pos 0 :cancel-state :not-cancellable}
    (progress/make (localization/message "a"))

    {:message (localization/message "b") :size 10 :pos 0 :cancel-state :not-cancellable}
    (progress/make (localization/message "b") 10)

    {:message (localization/message "c") :size 15 :pos 5 :cancel-state :not-cancellable}
    (progress/make (localization/message "c") 15 5)

    {:message (localization/message "d") :size 20 :pos 10 :cancel-state :not-cancellable}
    (progress/make (localization/message "d") 20 10 false)

    {:message (localization/message "e") :size 25 :pos 15 :cancel-state :cancellable}
    (progress/make (localization/message "e") 25 15 true)))

(deftest with-message-test
  (are [expected-fields progress]
    (let [actual-fields (to-map progress)]
      (is (instance? Progress progress))
      (is (= expected-fields actual-fields)))

    {:message (localization/message "new message") :size 20 :pos 10 :cancel-state :not-cancellable}
    (progress/with-message (progress/make (localization/message "d") 20 10 false) (localization/message "new message"))

    {:message (localization/message "other message") :size 25 :pos 15 :cancel-state :cancellable}
    (progress/with-message (progress/make (localization/message "e") 25 15 true) (localization/message "other message"))))

(deftest advance-test
  (are [expected-fields progress]
    (let [actual-fields (to-map progress)]
      (is (instance? Progress progress))
      (is (= expected-fields actual-fields)))

    {:message (localization/message "single step") :size 20 :pos 11 :cancel-state :not-cancellable}
    (progress/advance (progress/make (localization/message "single step") 20 10 false))

    {:message (localization/message "delta supplied") :size 25 :pos 20 :cancel-state :cancellable}
    (progress/advance (progress/make (localization/message "delta supplied") 25 15 true) 5)

    {:message (localization/message "message replaced") :size 25 :pos 20 :cancel-state :cancellable}
    (progress/advance (progress/make (localization/message "message") 25 15 true) 5 (localization/message "message replaced"))

    {:message (localization/message "past end") :size 2 :pos 2 :cancel-state :not-cancellable}
    (progress/advance (progress/make (localization/message "past end") 2 0 false) 5)

    {:message (localization/message "at end") :size 2 :pos 2 :cancel-state :not-cancellable}
    (progress/advance (progress/make (localization/message "at end") 2 2 false) 5)))

(deftest jump-test
  (testing "jump beyond size"
    (is (= 100 (progress/percentage (progress/jump (progress/make (localization/message "") 100) 666)))))
  (is (= 50 (progress/percentage (progress/jump (progress/make (localization/message "") 100 0) 50)))))

(deftest fraction-test
  (is (nil? (progress/fraction (progress/make (localization/message "mess") 0 0))))
  (is (= 0 (progress/fraction (progress/make (localization/message "mess")))))
  (is (= 1 (progress/fraction (progress/advance (progress/make (localization/message "mess"))))))
  (is (= 1/2 (progress/fraction (progress/advance (progress/make (localization/message "mess") 2))))))

(deftest percentage-test
  (is (nil? (progress/percentage (progress/make (localization/message "") 0 0))))
  (is (= 1 (progress/percentage (progress/make (localization/message "1%") 200 2))))
  (is (= 99 (progress/percentage (progress/make (localization/message "99%") 100 99)))))

(deftest message-test
  (is (= (localization/message "mess") (progress/message (progress/make (localization/message "mess"))))))

(deftest cancelable?-test
  (is (false? (progress/cancellable? (progress/make (localization/message "mess") 0 0 false))))
  (is (true? (progress/cancellable? (progress/make (localization/message "mess") 0 0 true))))
  (is (false? (progress/cancellable? (progress/cancel (progress/make (localization/message "mess") 0 0 true))))))

(deftest cancel-test
  (is (false? (progress/cancelled? (progress/make (localization/message "mess") 0 0 false))))
  (is (false? (progress/cancelled? (progress/make (localization/message "mess") 0 0 true))))
  (is (true? (progress/cancelled? (progress/cancel (progress/make (localization/message "mess") 0 0 true)))))
  (is (thrown? AssertionError (progress/cancel (progress/make (localization/message "mess") 0 0 false)))))

(deftest with-inherited-cancel-state-test
  (testing ":cancellable wins over :not-cancellable."
    (is (progress/cancellable?
          (progress/with-inherited-cancel-state
            (progress/->progress (localization/message "new") 1 0 :cancellable)
            (progress/->progress (localization/message "old") 1 0 :not-cancellable))))
    (is (progress/cancellable?
          (progress/with-inherited-cancel-state
            (progress/->progress (localization/message "new") 1 0 :not-cancellable)
            (progress/->progress (localization/message "old") 1 0 :cancellable)))))

  (testing ":cancelled wins over any state."
    (is (progress/cancelled?
          (progress/with-inherited-cancel-state
            (progress/->progress (localization/message "new") 1 0 :cancelled)
            (progress/->progress (localization/message "old") 1 0 :not-cancellable))))
    (is (progress/cancelled?
          (progress/with-inherited-cancel-state
            (progress/->progress (localization/message "new") 1 0 :cancelled)
            (progress/->progress (localization/message "old") 1 0 :cancellable))))
    (is (progress/cancelled?
          (progress/with-inherited-cancel-state
            (progress/->progress (localization/message "new") 1 0 :not-cancellable)
            (progress/->progress (localization/message "old") 1 0 :cancelled))))
    (is (progress/cancelled?
          (progress/with-inherited-cancel-state
            (progress/->progress (localization/message "new") 1 0 :cancellable)
            (progress/->progress (localization/message "old") 1 0 :cancelled)))))

  (testing "remaining field values are taken from the first Progress."
    (is (= {:message (localization/message "new message") :size 20 :pos 10 :cancel-state :not-cancellable}
           (to-map
             (progress/with-inherited-cancel-state
               (progress/->progress (localization/message "new message") 20 10 :not-cancellable)
               (progress/->progress (localization/message "old message") 1 0 :not-cancellable)))))
    (is (= {:message (localization/message "other message") :size 25 :pos 15 :cancel-state :not-cancellable}
           (to-map
             (progress/with-inherited-cancel-state
               (progress/->progress (localization/message "other message") 25 15 :not-cancellable)
               (progress/->progress (localization/message "old message") 1 0 :not-cancellable)))))))

(deftest nest-render-progress-test
  (let [last-progress (atom nil)
        render-progress! (fn [p] (reset! last-progress p))]
    (testing "plain nest"
      (let [nested-render-progress!
            (progress/nest-render-progress render-progress!
                                  (progress/make (localization/message "parent") 4 2))]
        ;; parent is at 2/4, child at 1/2 with the child
        ;; spanning 1 of parents progress
        ;; => effectively at 2/4 + 1 * 1/2 = 5/8
        (nested-render-progress! (progress/make (localization/message "child") 2 1))
        (is (= 5/8 (progress/fraction @last-progress)))))
    (testing "span nest"
      (let [nested-render-progress!
            (progress/nest-render-progress render-progress!
                                  (progress/make (localization/message "parent") 4 2)
                                  2)]
        ;; parent is at 2/4, child at 1/2 with the child
        ;; spanning 2 of parents progress
        ;; effectively at 2/4 + 2 * 1/2 = 3/4
        (nested-render-progress! (progress/make (localization/message "child") 2 1))
        (is (= 3/4 (progress/fraction @last-progress)))))
    (testing "precond failure"
      (is (thrown? AssertionError
                   (progress/nest-render-progress render-progress!
                                         (progress/make (localization/message "parent") 4 2)
                                         ;; span too large, 2 + 3 > 4
                                         3))))))

(deftest progress-mapv-test
  (let [render-res (atom [])
        render-f   #(swap! render-res conj %)
        message-f  #(localization/message (str "m" %))]

    (is (= (map inc (range 3))
           (progress/progress-mapv (fn [^long d _]
                                     (inc d))
                                   (range 3)
                                   render-f
                                   message-f)))
    (is (= [(progress/make (localization/message "m0") 3 0)
            (progress/make (localization/message "m1") 3 1)
            (progress/make (localization/message "m2") 3 2)]
           @render-res))

    (reset! render-res [])

    (testing "empty col"
      (is (= [] (progress/progress-mapv inc [] render-f)))
      (is (= [] @render-res)))

    (testing "nil"
      (is (= [] (progress/progress-mapv inc nil render-f)))
      (is (= [] @render-res)))))

(deftest test-done?
  (is (progress/done? progress/done))
  (is (not (progress/done? (progress/make (localization/message "job")))))
  (is (progress/done? (progress/advance (progress/make (localization/message "job")) 1)))
  (is (not (progress/done? (progress/advance (progress/make (localization/message "job") 2) 1))))
  (is (progress/done? (progress/make (localization/message "job") 2 2)))
  (is (not (progress/done? (progress/make (localization/message "job") 2 1)))))

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/ProgressTest.java
// Copyright 2020-2026 The Defold Foundation
// Copyright 2014-2020 King
// Copyright 2009-2014 Ragnar Svensson, Christian Murray
// Licensed under the Defold License version 1.0 (the "License"); you may not use
// this file except in compliance with the License.
//
// You may obtain a copy of the License, together with FAQs at
// https://www.defold.com/license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

package com.dynamo.bob;

import org.junit.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

import static org.junit.Assert.*;

public class ProgressTest {
    private record Report(IProgress.Message message, double fraction) {
    }

    private static final class RecordingReporter implements Progress.Reporter {
        private final List<Report> reports = new CopyOnWriteArrayList<>();
        private boolean canceled;
        private int closeCalls;

        @Override
        public void report(IProgress.Message message, double fraction) {
            reports.add(new Report(message, fraction));
        }

        @Override
        public void close() {
            closeCalls++;
        }

        @Override
        public boolean isCanceled() {
            return canceled;
        }
    }

    @Test
    public void messageReportsWithoutAdvancingProgress() {
        var reporter = new RecordingReporter();
        try (var progress = new Progress(reporter)) {
            progress.message(IProgress.Message.Building.INSTANCE);
        }

        assertEquals(2, reporter.reports.size());
        assertSame(IProgress.Message.Building.INSTANCE, reporter.reports.get(0).message());
        assertEquals(0.0, reporter.reports.get(0).fraction(), 0.0);
        assertSame(IProgress.Message.Building.INSTANCE, reporter.reports.get(1).message());
        assertEquals(1.0, reporter.reports.get(1).fraction(), 0.0);
    }

    @Test
    public void nonPositiveWorkDoesNotReport() {
        var reporter = new RecordingReporter();
        try (var progress = new Progress(reporter)) {
            var split = progress.split(3);
            split.worked(0);
            split.worked(-1);
        }

        assertEquals(1, reporter.reports.size());
        assertSame(IProgress.Message.Working.INSTANCE, reporter.reports.get(0).message());
        assertEquals(1.0, reporter.reports.get(0).fraction(), 0.0);
    }

    @Test
    public void repeatedWorkedCallsReachExactCompletionWithoutLeftovers() {
        var reporter = new RecordingReporter();
        try (var progress = new Progress(reporter)) {
            var split = progress.split(3);
            split.worked();
            split.worked();
            split.worked();
        }

        assertEquals(3, reporter.reports.size());
        assertSame(IProgress.Message.Working.INSTANCE, reporter.reports.get(0).message());
        assertEquals(1.0 / 3.0, reporter.reports.get(0).fraction(), 1e-15);
        assertSame(IProgress.Message.Working.INSTANCE, reporter.reports.get(1).message());
        assertEquals(2.0 / 3.0, reporter.reports.get(1).fraction(), 1e-15);
        assertSame(IProgress.Message.Working.INSTANCE, reporter.reports.get(2).message());
        assertEquals(1.0, reporter.reports.get(2).fraction(), 0.0);
    }

    @Test
    public void manyLazySubtasksAllowAnyTwoClosuresToReachCompletion() {
        var reporter = new RecordingReporter();
        try (var root = new Progress(reporter)) {
            var split = root.split(2);
            var subtasks = new ArrayList<IProgress>();
            for (int i = 0; i < 256; i++) {
                subtasks.add(split.subtask());
            }

            subtasks.get(17).close();
            subtasks.get(193).close();
            for (var subtask : subtasks) {
                subtask.close();
            }
        }

        assertEquals(2, reporter.reports.size());
        assertEquals(0.5, reporter.reports.get(0).fraction(), 1e-15);
        assertEquals(1.0, reporter.reports.get(1).fraction(), 0.0);
    }

    @Test
    public void workedAndSubtaskCallsShareTheSameLazySplit() {
        var reporter = new RecordingReporter();
        try (var root = new Progress(reporter)) {
            var split = root.split(4);
            split.worked();
            try (var child = split.subtask(2)) {
                child.close();
            }
            split.worked();
        }

        assertEquals(3, reporter.reports.size());
        assertEquals(0.25, reporter.reports.get(0).fraction(), 1e-15);
        assertEquals(0.75, reporter.reports.get(1).fraction(), 1e-15);
        assertEquals(1.0, reporter.reports.get(2).fraction(), 0.0);
    }

    @Test
    public void nestedSplitsAndSubtasksReachCompletionWithoutDuplicateFullReports() {
        var reporter = new RecordingReporter();
        try (var root = new Progress(reporter)) {
            var rootSplit = root.split(2);
            try (var first = rootSplit.subtask();
                 var second = rootSplit.subtask()) {
                var firstSplit = first.split(2);
                try (var grandchild = firstSplit.subtask()) {
                    grandchild.close();
                }
                firstSplit.worked();
                first.close();
                second.close();
            }
        }

        assertEquals(3, reporter.reports.size());
        assertEquals(0.25, reporter.reports.get(0).fraction(), 1e-15);
        assertEquals(0.5, reporter.reports.get(1).fraction(), 1e-15);
        assertEquals(1.0, reporter.reports.get(2).fraction(), 0.0);
        for (int i = 1; i < reporter.reports.size(); i++) {
            assertTrue(reporter.reports.get(i - 1).fraction() <= reporter.reports.get(i).fraction());
        }
    }

    @Test
    public void oversizedWorkAndSubtasksClampToWhatRemains() {
        var reporter = new RecordingReporter();
        try (var root = new Progress(reporter)) {
            var split = root.split(2);
            try (var child = split.subtask(Long.MAX_VALUE)) {
                child.close();
            }
            split.worked(Long.MAX_VALUE);
            split.worked(Long.MAX_VALUE);
        }

        assertEquals(1, reporter.reports.size());
        assertEquals(1.0, reporter.reports.get(0).fraction(), 0.0);
    }

    @Test
    public void isCanceledDelegatesToReporter() {
        var reporter = new RecordingReporter();
        try (var progress = new Progress(reporter)) {
            assertFalse(progress.isCanceled());
            reporter.canceled = true;
            assertTrue(progress.isCanceled());
        }
    }

    @Test
    public void closeReportsCompletionAndClosesReporter() {
        var reporter = new RecordingReporter();
        var progress = new Progress(reporter);
        var split = progress.split(4);

        progress.message(IProgress.Message.Building.INSTANCE);
        split.worked(1);
        progress.close();

        assertEquals(3, reporter.reports.size());
        assertSame(IProgress.Message.Building.INSTANCE, reporter.reports.get(2).message());
        assertEquals(1.0, reporter.reports.get(2).fraction(), 0.0);
        assertEquals(1, reporter.closeCalls);
        assertFalse(progress.isCanceled());
    }

    @Test
    public void closeStopsFurtherNotifications() {
        var reporter = new RecordingReporter();
        var progress = new Progress(reporter);

        progress.close();
        progress.message(IProgress.Message.Cleaning.INSTANCE);
        progress.close();

        assertEquals(1, reporter.reports.size());
        assertSame(IProgress.Message.Working.INSTANCE, reporter.reports.get(0).message());
        assertEquals(1.0, reporter.reports.get(0).fraction(), 0.0);
        assertEquals(1, reporter.closeCalls);
    }

}
